store/src/java/com/zimbra/cs/html/OwaspHtmlSanitizer.java

```diff
@@ -1,36 +1,409 @@
- * ***** END LICENSE BLOCK *****
- */
-
 package com.zimbra.cs.html;
 
-import java.io.IOException;
-import java.io.InputStream;
-import java.io.Reader;
-import java.io.Writer;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.regex.Pattern;
 
-public class OwaspHtmlSanitizer extends AbstractDefang {
+import org.apache.commons.lang.StringEscapeUtils;
+import org.owasp.html.AttributePolicy;
+import org.owasp.html.CssSchema;
+import org.owasp.html.ElementPolicy;
+import org.owasp.html.Handler;
+import org.owasp.html.HtmlPolicyBuilder;
+import org.owasp.html.HtmlSanitizer;
+import org.owasp.html.HtmlSanitizer.Policy;
+import org.owasp.html.HtmlStreamRenderer;
+import org.owasp.html.PolicyFactory;
 
-    @Override
-    public void defang(InputStream is, boolean neuterImages, Writer out) throws IOException {
-        // TODO to be implemented later
+import com.google.common.collect.ImmutableSet;
+import com.zimbra.common.localconfig.DebugConfig;
+import com.zimbra.common.util.StringUtil;
+import com.zimbra.common.util.ZimbraLog;
+import com.zimbra.cs.servlet.ZThreadLocal;
+
+class OwaspHtmlSanitizer implements Callable<String> {
+
+    /** The Host header received in the request. */
+    private String reqVirtualHost = null;
+    /** enable same host post request for a form in email */
+    private static boolean sameHostFormPostCheck = DebugConfig.defang_block_form_same_host_post_req;
+    // regexes inside of attr values to strip out
+    private static final Pattern AV_JS_ENTITY = Pattern.compile(DebugConfig.defangAvJsEntity);
+    private static final Pattern AV_SCRIPT_TAG = Pattern.compile(DebugConfig.defangAvScriptTag,
+        Pattern.CASE_INSENSITIVE);
+    private static final Pattern VALID_IMG_FILE = Pattern.compile(DebugConfig.defangValidImgFile);
+    private static final Pattern VALID_INT_IMG = Pattern.compile(DebugConfig.defangValidIntImg,
+        Pattern.CASE_INSENSITIVE);
+    private static final Pattern VALID_EXT_URL = Pattern.compile(DebugConfig.defangValidExtUrl,
+        Pattern.CASE_INSENSITIVE);
+    // matches the file format that convertd uses so it doesn't get removed
+    private static final Pattern VALID_CONVERTD_FILE = Pattern
+        .compile(DebugConfig.defangValidConvertdFile);
+
+    public OwaspHtmlSanitizer(String html, boolean neuterImages) {
+        this.html = html;
+        this.neuterImages = neuterImages;
+        if (ZThreadLocal.getRequestContext() != null) {
+            this.reqVirtualHost = ZThreadLocal.getRequestContext().getVirtualHost();
+        }
     }
 
+    /**
+     * The following CSS properties do not appear in the default whitelist from
+     * OWASP, but they improve the fidelity of the HTML display without
+     * unacceptable risk.
+     */
+    private static final CssSchema ADDITIONAL_CSS = CssSchema
+        .withProperties(ImmutableSet.of("float", "display"));
```

Comment:
FWIW, allowing for the ability to do something like 'display: none' is considered a security risk, especially in the right context.  In an ideal world, I don't think that would be considered an acceptable risk.